drivers/sensor/adltc2990/adltc2990.c

```diff
@@ -0,0 +1,485 @@
+	case ADLTC2990_MODE_V1_MINUS_V2_V3_MINUS_V4: {
+		type = VOLTAGE_DIFFERENTIAL;
+		break;
+	}
+	case ADLTC2990_MODE_TR1_V3_V4:
+	case ADLTC2990_MODE_TR1_V3_MINUS_V4: {
+	case ADLTC2990_MODE_TR1_TR2:
+		type = TEMPERATURE;
+		break;
+	}
+	default: {
+		break;
+	}
+	}
+	return type;
+}
+
+static enum adltc2990_monitoring_type adltc2990_get_v3_v4_measurement_modes(uint8_t mode_4_3,
+									    uint8_t mode_2_0)
+{
+	if (mode_2_0 > ADLTC2990_MODE_2_0_MAX_VALUE || mode_4_3 > ADLTC2990_MODE_4_3_MAX_VALUE) {
+		LOG_ERR("Invalid Measurement Mode");
+		return -EINVAL;
+	}
+	if (mode_4_3 == ADLTC2990_MEASURE_INTERNAL_TEMPERATURE_ONLY ||
+	    mode_4_3 == ADLTC2990_MEASURE_PINS_V1_V2_ONLY) {
+		return NOTHING;
+	}
+
+	enum adltc2990_monitoring_type type = NOTHING;
+
+	switch (mode_2_0) {
+	case ADLTC2990_MODE_V1_V2_TR2:
+	case ADLTC2990_MODE_V1_MINUS_V2_TR2:
+	case ADLTC2990_MODE_TR1_TR2: {
+		type = TEMPERATURE;
+		break;
+	}
+
+	case ADLTC2990_MODE_V1_MINUS_V2_V3_V4:
+	case ADLTC2990_MODE_TR1_V3_V4:
+	case ADLTC2990_MODE_V1_V2_V3_V4: {
+		type = VOLTAGE_SINGLEENDED;
+		break;
+	}
+	case ADLTC2990_MODE_TR1_V3_MINUS_V4:
+	case ADLTC2990_MODE_V1_MINUS_V2_V3_MINUS_V4: {
+		type = VOLTAGE_DIFFERENTIAL;
+		break;
+	}
+	default: {
+		break;
+	}
+	}
+	return type;
+}
+
+static bool adltc2990_is_busy(const struct device *dev)
+{
+	const struct adltc2990_config *cfg = dev->config;
+	uint8_t status_reg = 0;
+
+	i2c_reg_read_byte_dt(&cfg->bus, ADLTC2990_REG_STATUS, &status_reg);
+	return status_reg & BIT(0);
+}
+
+static void adltc2990_get_v1_v2_val(const struct device *dev, struct sensor_value *val,
+				    uint8_t num_values, uint8_t *const offset_index)
+{
+	struct adltc2990_data *data = dev->data;
+
+	for (uint8_t index = 0; index < num_values; index++) {
+		val[index].val1 = data->pins_v1_v2_values[index] / 1000000;
+		val[index].val2 = data->pins_v1_v2_values[index] % 1000000;
+		*offset_index = index + 1;
+	}
+}
+
+static void adltc2990_get_v3_v4_val(const struct device *dev, struct sensor_value *val,
+				    uint8_t num_values, uint8_t const *const offset)
+{
+	struct adltc2990_data *data = dev->data;
+
+	uint8_t offset_index = *offset;
+
+	for (uint8_t index = 0; index < num_values; index++) {
+		val[index + offset_index].val1 = data->pins_v3_v4_values[index] / 1000000;
+		val[index + offset_index].val2 = data->pins_v3_v4_values[index] % 1000000;
+	}
+}
+
+static int adltc2990_trigger_measurement(const struct device *dev)
+{
+	const struct adltc2990_config *cfg = dev->config;
+
+	return i2c_reg_write_byte_dt(&cfg->bus, ADLTC2990_REG_TRIGGER, 0x1);
+}
+
+static int32_t adtc2990_raw_property_value(uint8_t msb_value, uint8_t lsb_value, uint16_t bitmask,
+					   uint8_t negative_bit_index)
+{
+	const uint16_t msb = msb_value << 8U;
+
+	const uint16_t lsb = lsb_value;
+
+	const uint16_t val = msb + lsb;
+
+	const bool is_neg = (msb_value & negative_bit_index) ? true : false;
+	int32_t ret = is_neg ? ((~val & bitmask) + 1) * -1 : (val & bitmask);
```

Comment:
Is this trying to sign extend? It would be simpler to left shift then right shift, and you can do that unconditionally without checking if the quantity is negative.